mkckks/evaluator.go

```diff
@@ -0,0 +1,472 @@
+	}
+
+	// finalize computation of c0'
+	index := ring.PermuteNTTIndex(galEl, eval.ringQP.N)
+	ring.PermuteNTTWithIndexLvl(level, c.Ciphertexts.Value[0], index, res[0])
+
+	tmpModDown := eval.ringQ.NewPoly()
+	eval.convertor.ModDownSplitNTTPQ(level, restmpQ[0], restmpP[0], tmpModDown)
+	eval.ringQ.AddLvl(level, res[0], tmpModDown, res[0])
+
+	// finalize computation of ci'
+	for i := uint64(1); i <= k; i++ {
+
+		eval.convertor.ModDownSplitNTTPQ(level, restmpQ[i], restmpP[i], tmpModDown)
+		eval.ringQ.CopyLvl(level, tmpModDown, res[i])
+	}
+
+	out.Ciphertexts.SetValue(res)
+
+	return out
+}
+
+// prepare rotation keys for operations in split crt basis
+func prepareRotationKey(j, level, modulus, beta uint64, rotKeys []*mkrlwe.MKRotationKey, gal0QP, gal1QP *rlwe.SwitchingKey) {
+
+	for u := uint64(0); u < beta; u++ {
+		gal0QP.Value[u][0].Coeffs = rotKeys[j-1].Key.Value[u][0].Coeffs[:level+1]
+		gal0QP.Value[u][1].Coeffs = rotKeys[j-1].Key.Value[u][0].Coeffs[modulus:]
+		gal1QP.Value[u][0].Coeffs = rotKeys[j-1].Key.Value[u][1].Coeffs[:level+1]
+		gal1QP.Value[u][1].Coeffs = rotKeys[j-1].Key.Value[u][1].Coeffs[modulus:]
+	}
+
+}
+
+// NewPlaintextFromValue returns a plaintext from the provided values
+func (eval *mkEvaluator) NewPlaintextFromValue(values []complex128) *ckks.Plaintext {
+
+	return eval.encoder.EncodeNTTAtLvlNew(eval.params.MaxLevel(), values, eval.params.LogSlots())
+}
+
+func getCiphertextIndex(peerID uint64, ct *MKCiphertext) uint64 {
+
+	for i, id := range ct.PeerID {
+
+		if id == peerID {
+			return uint64(i + 1)
+		}
+	}
+
+	return 0
+}
+
+func checkParticipantsRelinKey(peerID []uint64, relinKeys []*mkrlwe.MKRelinearizationKey) {
+
+	for i, id := range peerID {
+		if id != relinKeys[i].PeerID {
+			panic("Incorrect evaluation keys for the given ciphertexts")
+		}
+	}
+}
+
+func checkParticipantsPubKey(peerID []uint64, pubKeys []*mkrlwe.MKPublicKey) {
+
+	for i, id := range peerID {
+		if id != pubKeys[i].PeerID {
+			panic("Incorrect public keys for the given ciphertexts")
+		}
+	}
+}
+
+func checkParticipantsRotKey(peerID []uint64, rotKeys []*mkrlwe.MKRotationKey) {
+
+	for i, id := range peerID {
+		if id != rotKeys[i].PeerID {
+			panic("Incorrect galois evaluation keys for the given ciphertexts")
+		}
+	}
+}
+
+// evaluateInPlace performs an evaluation of the evaluate function between c0 and c1, and stores the result into ctOut by optimizing the computation if there are nils. Subtraction needs to be handled ad-hoc and is indicated with isSub bool
+func (eval *mkEvaluator) evaluateInPlace(c0, c1, ctOut *ckks.Ciphertext, isSub bool, evaluate func(uint64, *ring.Poly, *ring.Poly, *ring.Poly)) {
+
+	var tmp0, tmp1 *ckks.Element
+	level := utils.MinUint64(utils.MinUint64(c0.Level(), c1.Level()), ctOut.Level())
+
+	minDegree := utils.MinUint64(c0.Degree(), c1.Degree())
+
+	if c1.Scale() > c0.Scale() {
+		tmp0 = new(ckks.Element)
```

Comment:
I have been using the multi-key functionality implemented here for CKKS for the last two days and I think there is a problem with this allocation. In my example, I had two ciphertexts with different scales that were added together and this would throw errors. The error is thrown when you try to access the Level() on the passed object created here and it's thrown because the Value array from the element is empty and the code tries to access the element in position 0. My solution was to call `MultByConstNew` in the next if statement and copy the value to tmp0, but maybe the tmp0 initialisation can be changed.

Reply:
Hello. You're completely right. The same problem should arise for the tmp1 variable at line 440. In the `ckks.evaluator`, their implementation of `evaluateInPlace` uses a ciphertext pool with pre initialized ciphertexts so it can be fixed that way. Else, calling `MultByConstNew` will effectively correct the bug.

Maybe you could ask Christian or Jean-Phillipe what would be the best option.

Reply:
Hello @andru47.
Thanks for reporting this finding. We'll fix this bug soon as we will soon try to merge this branch with the master branch.